livekit-agents/livekit/agents/voice/agent.py

```
@@ -70,15 +70,17 @@ def tools(self) -> list[llm.FunctionTool]:
     @property
     def chat_ctx(self) -> llm.ChatContext:
         """
-        Provides a read-only view of the agent's current chat context.
-
         Returns:
-            llm.ChatContext: A read-only version of the agent's conversation history.
+            llm.ChatContext: A read-only version of the agent's conversation history if the agent's
+            LLM is a RealtimeModel. Otherwise, it returns the internal mutable chat context.
 
         See Also:
             update_chat_ctx: Method to update the internal chat context.
         """
-        return _ReadOnlyChatContext(self._chat_ctx.items)
+        if isinstance(self.llm, llm.RealtimeModel):
+            return _ReadOnlyChatContext(self._chat_ctx.items)
+        else:
+            return self._chat_ctx
```

@theomonnom wdyt about make `agent.chat_ctx` mutable for regular LLM?

---

I understand why we want to do it, but I would not create major differences between both today. This may be very confusing when switching an agent using RealtimeLLM to a stateless LLM. The code will break

---

I found it's a little bit annoying and inefficient if we make `chat_ctx` of regular LLM readonly, like in this example https://github.com/livekit/agents/pull/1695/files#diff-88ec3d07b4e65630ae21331f8576f5f677e9d5ad4a52e0e8d15c33b573b81846R35-R36

user should now what LLM is using in an agent for most of the case, or they can use `chat_ctx.readonly` flag to use different methods if it's for both RealtimeModel and LLM.

```python
        chat_ctx = chat_ctx.copy()
        chat_ctx.items.append(new_message)
        self.update_chat_ctx(chat_ctx)
        # we copied the chat_ctx 3 times here
 
        # or 
        assert chat_ctx.mutable  # asset not chat_ctx.readonly
        chat_ctx.items.append(new_message)
```

But I don't have a strong opinion here, maybe we wait for the future. But I think there will be a lot questions about that, @lukasIO just met this issue that modifying the readonly chat_ctx failed silently because it was in a async method... So maybe at least it's better to show the error message in a log then raise the error.

---

but anyway, raise the error is better than allow it silently, which is much hard to find lol

---

how are users expected to use `conversation_item_added` event effectively when they need to reconstruct the whole chat ctx each time they update it 🤔

---

Where was the async method that was failing silently? We should absolutely log all errors from tasks we create. 

I think I would say to see that later, my opinion is that a code you see on our docs/github should directly be compatible with what you have locally, no matter what model you're using.

I think as a first step, we should maybe allow update_chat_ctx to not be necessarily `awaited`. But I think it's fine for v1.0? 

I wouldn't worry about the efficiency here, we only do shallow copies, it's very fast

---

I think a "blessed" way to hook into the chat ctx for a persistence layer should not be an afterthought but rather an integral part of 1.0. 

If it's not the `conversation_item_added` event, it's fine for it to be a custom text sink, but if that's the case, then we need all necessary bits exposed publicly

---

As I understand the main concern is a difference in behaviour between realtime model and regular llm. 

I guess we could just abstract that away for realtime models? where for realtime models a call to `add_message` would update the chat context with the copy? 
That seems more obvious than requiring to copy + update the chat ctx for all models in user land

---

It may failed in user defined async function without log. But anyway I'll revert it for now to unblock this pr.